spacy/tokenizer.pyx

```diff
@@ -174,23 +211,124 @@ cdef class Tokenizer:
             yield self(text)
 
     def _flush_cache(self):
-        self._reset_cache([key for key in self._cache if not key in self._specials])
+        self._reset_cache([key for key in self._cache])
 
     def _reset_cache(self, keys):
         for k in keys:
+            cached = <_Cached*>self._cache.get(k)
             del self._cache[k]
-            if not k in self._specials:
-                cached = <_Cached*>self._cache.get(k)
-                if cached is not NULL:
-                    self.mem.free(cached)
+            if cached is not NULL:
+                self.mem.free(cached)
 
-    def _reset_specials(self):
+    def _flush_specials(self):
         for k in self._specials:
             cached = <_Cached*>self._specials.get(k)
             del self._specials[k]
             if cached is not NULL:
                 self.mem.free(cached)
 
+    cdef int _apply_special_cases(self, Doc doc):
+        """Retokenize doc according to special cases.
+
+        doc (Doc): Document.
+        """
+        cdef int i
+        cdef int j
+        cdef int curr_length = doc.length
+        cdef int max_length = 0
+        cdef int offset = 0
+        cdef int span_length_diff = 0
+        cdef bint modify_in_place = True
+        cdef int idx_offset = 0
+        cdef int orig_final_spacy
+        cdef int orig_idx
+        cdef Pool mem = Pool()
+        cdef vector[MatchStruct] c_matches
+        self._special_matcher.find_matches(doc, &c_matches)
+        # Skip processing if no matches
+        if c_matches.size() == 0:
+            return True
+        cdef vector[MatchStruct] c_filtered
+        self._filter_spans(c_matches, c_filtered, doc.length)
+        spans = [doc[match.start:match.end] for match in c_filtered]
+        # Put span info in span.start-indexed dict and calculate maximum
+        # intermediate document size
+        span_data = {}
+        for span in spans:
+            rule = self._rules.get(span.text, None)
+            span_length_diff = 0
+            if rule:
+                span_length_diff = len(rule) - (span.end - span.start)
+            if span_length_diff > 0:
+                modify_in_place = False
+            curr_length += span_length_diff
+            if curr_length > max_length:
+                max_length = curr_length
+            span_data[span.start] = (span.text, span.start, span.end, span_length_diff)
+        # Modify tokenization according to filtered special cases
+        # If modifications never increase doc length, can modify in place
+        if modify_in_place:
```

Comment:
Alternatively, do without `modify_in_place` variable and just check `curr_length == doc.length`